tests/system/Validation/RulesTest.php

```diff
@@ -614,6 +614,33 @@ public function requiredWithProvider(): Generator
         ];
     }
 
+    /**
+     * @dataProvider RequiredWithAndOtherRuleProvider
+     */
+    public function testRequiredWithAndOtherRule(bool $expected, array $data): void
+    {
+        $this->validation->setRules([
+            'testField' => 'required_with[otherField]|valid_date',
```

Comment:
If you run this test with: `required_with[otherField]|permit_empty|valid_date` it will still pass in the current dev branch.

There are two explanations: the test doesn't cover the original problem or the problem doesn't exist.

Reply:
Exactly. Maybe my understanding of `permit_empty` is wrong.
But the rule `required_with[otherField]|permit_empty` seems odd to me.
Required and permit empty? What's that?

@shishamo Can you show a test case with `permit_empty` that fails in the current `develop` branch?

Reply:
```php
$validation = Services::validation();
$validation = $validation->setRules([
    'mustBeADate' => 'required_with[otherField]|permit_empty|valid_date',
]);

$data = [
    'otherField' => 'exists',
    'mustBeADate' => [],
];
$validation->run($data); // return true and must return false

$data = [
    'otherField' => 'exists',
    'mustBeADate' => null,
];
$validation->run($data); // return true and must return false

$data = [
    'otherField' => 'exists',
    'mustBeADate' => '',
];
$validation->run($data); // return true and must return false
```

Reply:
@kenjis 
this 3 previous case are wrong, and i think than they should return false
in my case, `mustBeADate` must be a required correct date format only in `otherField` exists

Reply:
@shishamo Thank you.

But the following test passes in `develop`:
```php
    /**
     * @dataProvider RequiredWithAndOtherRuleProvider
     */
    public function testRequiredWithAndOtherRule(bool $expected, array $data): void
    {
        $this->validation->setRules([
            'mustBeADate' => 'required_with[otherField]|permit_empty|valid_date',
        ]);

        $result = $this->validation->run($data);

        $this->assertSame($expected, $result);
    }

    public function RequiredWithAndOtherRuleProvider(): Generator
    {
        yield from [
            [true, []],
            [true, ['mustBeADate' => '2023-06-12']],
            [true, ['mustBeADate' => '2023-06-12', 'otherField' => 'exists']],
            [true, ['mustBeADate' => '2023-06-12', 'otherField' => '']],
            [true, ['mustBeADate' => '', 'otherField' => '']],
            [false, ['mustBeADate' => '', 'otherField' => 'exists']],
            [false, ['mustBeADate' => [], 'otherField' => 'exists']],
            [false, ['mustBeADate' => null, 'otherField' => 'exists']],
            [false, ['otherField' => 'yes']],
        ];
    }
```

Reply:
sorry, i made a mistake, i checks and it was working in that way.

the problem is if the `otherField` does not exist and the client try to send a value 'mustBeADate' = [] or 'mustBeADate' = null or 'mustBeADate' = '', the data will pass

```php
yield from [
    [false, ['mustBeADate' => ''],
    [false, ['mustBeADate' => null]],
    [false, ['mustBeADate' => []],
];
```

i am not sure but should these case pass?
if these case pass it will might allow the client to pass some unexpected values i think

Reply:
it is just an idea, but maybe but if `otherField` does not exit BUT `mustBeADate` exists the validations (`valid_date` in that case and other validation if has) should run?

what do you think?

Reply:
@michalsn Is the following interpretation correct?

> `required_with[otherField]|permit_empty|valid_date`

1. When `otherField` exists, it means `required|permit_empty|valid_date`
   - `required` has more priority, so it is `required|valid_date`
   -   empty is not permitted 
2. When `otherField` does not exist, `permit_empty|valid_date`
   - empty is permitted 

> `required_with[otherField]|valid_date`

1. When `otherField` exists, it means `required|valid_date`
   - empty is not permitted 
3. When `otherField` does not exist, `valid_date`
   - empty is not permitted

Reply:
> it is just an idea, but maybe if otherField does not exit BUT mustBeADate exists the validations (valid_date in that case and other validation if has) should run?

Should run. If not run, the value is not validated at all.

Reply:
@shishamo 
> i am not sure but should these case pass?
> if these case pass it will might allow the client to pass some unexpected values i think

Should pass because you permit empty. These values are all empty in the `permit_empty` rule.
When you permit empty, empty values are expected.

If you don't permit empty,`mustBeADate` must exist and be always a valid date, not an empty.

@michalsn 
> It is not easy to do different combinations, but no one said that our built-in rules should handle all cases.

Yeah, it seems not possible to handle this case.

Reply:
I understand 

Just a question then:
What is required_with rule for if I cannot apply rules only if another field exist?

Reply:
@kenjis 

> @michalsn Is the following interpretation correct?
> 
> > `required_with[otherField]|permit_empty|valid_date`
> 
> 1. When `otherField` exists, it means `required|permit_empty|valid_date`
>    
>    * `required` has more priority, so it is `required|valid_date`
>    * empty is not permitted
> 2. When `otherField` does not exist, `permit_empty|valid_date`
>    
>    * empty is permitted

Yes.

> > `required_with[otherField]|valid_date`
> 
> 1. When `otherField` exists, it means `required|valid_date`
>    
>    * empty is not permitted
> 2. When `otherField` does not exist, `valid_date`
>    
>    * empty is not permitted

Yes. But in general, it's redundant. We don't need `required_with` (or `required`), because the `valid_date` rule causes the value to be required.

Reply:
@shishamo 

The main issue here is that `permit_empty` rule is too wide for your use case. From what you're saying you would need something like `permit_not_exists` rule.  

For now, I would try to write a separate rule to handle your case or try to add a rule conditionally:

```php
if ($this->request->getPost('otherField') !== null) {
        $validation->setRule('mustBeADate', 'Date', 'valid_date');
}
```

Getting the validated data will cause some more work but in 4.4 handling this will be much easier since we will have `$validation->getValidated()` method, which will return only validated data.

> What is required_with rule for if I cannot apply rules only if another field exist?

To require some fields conditionally. Example:

```php
$validation->setRules([
        'married' => ['rules' => ['in_list[0,1]']],
        'partner_name' => ['rules' => ['required_with[married]']],
 ]);
```

Reply:
@michalsn Thank you!
I did not realize `required` has higher priority than `permit_empty`.
So I misunderstood the behavior.

Reply:
```php
$validation->setRules([
        'married' => ['rules' => ['in_list[0,1]']],
        'partner_name' => ['rules' => ['required_with[married]']],
 ]);
```
This code works.

I misunderstood the behavior. I sent a PR to fix the docs:
https://github.com/codeigniter4/CodeIgniter4/pull/7558/files#diff-75e5e834c9c260f8111e1ebc2fcab1b523eb040c6ee1ac31f9a65f4895a9a0f0L819-R823